Authenticators/IMAP/imapauth.py

```diff
@@ -0,0 +1,605 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8
+
+# Based on work "smfauth" which is Copyright (C) 2010 Stefan Hacker <dd0t@users.sourceforge.net> 
+# https://raw.githubusercontent.com/mumble-voip/mumble-scripts/master/Authenticators/SMF/2.0/smfauth.py
+# 
+# Auth for IMAP and python3 compatibility by dadosch, 2018
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+
+# - Redistributions of source code must retain the above copyright notice,
+#   this list of conditions and the following disclaimer.
+# - Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+# - Neither the name of the Mumble Developers nor the names of its
+#   contributors may be used to endorse or promote products derived from this
+#   software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# `AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#
+#    imapauth.py - Authenticator implementation for password authenticating
+#                 a Murmur server against an IMAP server
+#
+#    Requirements:
+#        * python >=3 and the following python modules:
+#            * ice-python
+#            * IMAPClient
+#            * daemon (when run as a daemon)
+#
+
+import sys
+import Ice
+import _thread
+import logging
+import configparser
+
+from threading  import Timer
+from optparse   import OptionParser
+from logging    import (debug,
+                        info,
+                        warning,
+                        error,
+                        critical,
+                        exception,
+                        getLogger)
+
+
+import hashlib
+import ssl
+from imapclient import IMAPClient
+
+ssl_context = ssl.create_default_context()
+
+def hashString(s):
+    """Helper function to convert strings to a fixed integer"""
+    result = 0
+    hash = hashlib.md5(s.encode('utf8')).digest()
```

Comment:
There are probably better hash algorithms with a severely less chance of collisions. I’d prefer one of those.

Reply:
Which would be one of those? Remember that the user id must not be too long, otherwise it won't work (see line 369)

Reply:
A sha algorithm for example. Not sure what python provides in its standard lib.

But I’m not even sure how the conversion/reduction to int will interact with that. Without doing some risk analysis/probability reasoning I think it’s probably better to introduce an explicit user mapping.

I am willing to accept this with a strong warning in the description/header.

I’d feel more comfortable with some idea of how likely collisions are though. If it is very unlikely the warning will not have to be as strongly worded.